### PR TITLE
[Mellanox] Use MAC from EEPROM for PortChannels and VLAN Interfaces

### PR DESCRIPTION
**Pull request by andriymoroz-mlnx**
Signed-off-by: Andriy Moroz <c_andriym@mellanox.com>

**- What I did**
Updated teams start script to pass MAC from EEPROM to the portchannels config template

**- How I did it**
updated start.sh which runs in teamd container 

**- How to verify it**
Start SONiC and make sure MACs on PortChannel (and members) or Vlan interface are similar to one returned by command
od -vt x1 -An /sys/bus/i2c/devices/8-0051/eeprom | xargs printf "0x%s " | xargs printf "%02x:" | awk 'BEGIN { FS=":"; i=8+1+2+1} {while(i<NF) {type=$i; len=("0x"$(i+1));if(type!="24") {i=i+2+len} else {print substr($0, (i+1)*3+1, len*3-1); break}}}'

some bits in last byte may differ

### CODE REVIEW

**Comment by andriymoroz-mlnx** [comment]
need to do some tests to implement the same for VLAN interfaces...

File: dockers/docker-teamd/start.sh [hunk 1]
```diff
@@ -6,7 +6,12 @@ rm -rf $TEAMD_CONF_PATH
 mkdir -p $TEAMD_CONF_PATH
 
 SONIC_ASIC_TYPE=$(sonic-cfggen -y /etc/sonic/sonic_version.yml -v asic_type)
-MAC_ADDRESS=$(ip link show eth0 | grep ether | awk '{print $2}')
+
+if [ "$SONIC_ASIC_TYPE" == "mellanox" ]; then
+    MAC_ADDRESS=$(od -vt x1 -An /sys/bus/i2c/devices/8-0051/eeprom | xargs printf "0x%s " | xargs printf "%02x:" | awk 'BEGIN { FS=":"; i=8+1+2+1} {while(i<NF) {type=$i; len=("0x"$(i+1));if(type!="24") {i=i+2+len} else {print substr($0, (i+1)*3+1, len*3-1); break}}}')
+else
+    MAC_ADDRESS=$(ip link show eth0 | grep ether | awk '{print $2}')
```
Comment (pavel-shirshov) [inline dockers/docker-teamd/start.sh]: cat /sys/class/net/eth0/address

File: dockers/docker-teamd/start.sh [hunk 2]
```diff
@@ -6,7 +6,12 @@ rm -rf $TEAMD_CONF_PATH
 mkdir -p $TEAMD_CONF_PATH
 
 SONIC_ASIC_TYPE=$(sonic-cfggen -y /etc/sonic/sonic_version.yml -v asic_type)
-MAC_ADDRESS=$(ip link show eth0 | grep ether | awk '{print $2}')
+
+if [ "$SONIC_ASIC_TYPE" == "mellanox" ]; then
+    MAC_ADDRESS=$(od -vt x1 -An /sys/bus/i2c/devices/8-0051/eeprom | xargs printf "0x%s " | xargs printf "%02x:" | awk 'BEGIN { FS=":"; i=8+1+2+1} {while(i<NF) {type=$i; len=("0x"$(i+1));if(type!="24") {i=i+2+len} else {print substr($0, (i+1)*3+1, len*3-1); break}}}')
```
Comment (lguohan) [inline dockers/docker-teamd/start.sh]: sudo decode-syseeprom -m to figure out the base mac?

Comment (andriymoroz-mlnx) [reply dockers/docker-teamd/start.sh]: it is not available inside docker container, checking why...

File: src/sonic-config-engine/sonic_platform.py [hunk 3]
```diff
@@ -41,12 +41,19 @@ def get_sonic_version_info():
     return data
 
 def get_system_mac():
-    proc = subprocess.Popen("ip link show eth0 | grep ether | awk '{print $2}'", shell=True, stdout=subprocess.PIPE)
+    get_mac_cmd = "ip link show eth0 | grep ether | awk '{print $2}'"
+    version_info = get_sonic_version_info()
+
+    if (version_info['asic_type'] == 'mellanox'):
+        eeprom_file = "/sys/bus/i2c/devices/8-0051/eeprom"
+        if (os.access(eeprom_file, os.R_OK)):
```
Comment (qiluo-msft) [inline src/sonic-config-engine/sonic_platform.py]: >if (os.access(eeprom_file, os.R_OK)): [](start = 8, length = 37)

else, it is an error.

File: files/build_templates/updategraph.service.j2 [hunk 4]
```diff
@@ -6,6 +6,10 @@ Requires=database.service
 
 [Service]
 Type=oneshot
+{% if sonic_asic_platform == 'mellanox' -%}
+EnvironmentFile=/host/machine.conf
+ExecStartPre=/bin/bash -c "/usr/share/sonic/device/$onie_platform/hw-management start"
```
Comment (qiluo-msft) [inline files/build_templates/updategraph.service.j2]: >hw-management [](start = 66, length = 13)

Is updategraph a good location for hw-management?

Comment (taoyl-ms) [reply files/build_templates/updategraph.service.j2]: Agree with Qi. I don't see the reason to put hw-management here, and I don't see the need to change this into a j2 file either. Platform information can also be tested during runtime.

Comment (andriymoroz) [reply files/build_templates/updategraph.service.j2]: we need drivers to access eeprom
j2 to avoid this code for others

Comment (taoyl-ms) [reply files/build_templates/updategraph.service.j2]: Why not create another service for this?
And, you can create a shell script to enable hw-management in which you can test platform information run time. It is much easier to follow than a build-time j2 rendering.

Comment (lguohan) [reply files/build_templates/updategraph.service.j2]: in current case, when will hw-management be started? Maybe hw-management itself should be a service and make it start early.

File: dockers/docker-teamd/start.sh [hunk 5]
```diff
@@ -6,7 +6,12 @@ rm -rf $TEAMD_CONF_PATH
 mkdir -p $TEAMD_CONF_PATH
 
 SONIC_ASIC_TYPE=$(sonic-cfggen -y /etc/sonic/sonic_version.yml -v asic_type)
-MAC_ADDRESS=$(ip link show eth0 | grep ether | awk '{print $2}')
+
+if [ "$SONIC_ASIC_TYPE" == "mellanox" ]; then
+    MAC_ADDRESS=$(sonic-cfggen -d -v DEVICE_METADATA.localhost.mac)
+else
+    MAC_ADDRESS=$(ip link show eth0 | grep ether | awk '{print $2}')
```
Comment (pavel-shirshov) [inline dockers/docker-teamd/start.sh]: MAC_ADDRESS=$(</sys/class/net/eth0/address)

File: files/build_templates/sonic_debian_extension.j2 [hunk 6]
```diff
@@ -155,7 +155,7 @@ sudo LANG=C chroot $FILESYSTEM_ROOT systemctl enable hostname-config.service
 sudo cp $IMAGE_CONFIGS/hostname/hostname-config.sh $FILESYSTEM_ROOT/usr/bin/
 
 # Copy updategraph script and service file
-sudo cp $IMAGE_CONFIGS/updategraph/updategraph.service  $FILESYSTEM_ROOT/etc/systemd/system/
+j2 files/build_templates/updategraph.service.j2 | sudo tee $FILESYSTEM_ROOT/etc/systemd/system/updategraph.service
```
Comment (pavel-shirshov) [inline files/build_templates/sonic_debian_extension.j2]: why do we need to copy everything to the screen?

File: src/sonic-config-engine/sonic_platform.py [hunk 7]
```diff
@@ -41,12 +41,18 @@ def get_sonic_version_info():
     return data
 
 def get_system_mac():
-    proc = subprocess.Popen("ip link show eth0 | grep ether | awk '{print $2}'", shell=True, stdout=subprocess.PIPE)
+    version_info = get_sonic_version_info()
+
+    if (version_info['asic_type'] == 'mellanox'):
+        get_mac_cmd = "sudo decode-syseeprom -m"
+    else:
+        get_mac_cmd = "ip link show eth0 | grep ether | awk '{print $2}'"
```
Comment (pavel-shirshov) [inline src/sonic-config-engine/sonic_platform.py]: read a file "/sys/class/net/eth0/address"?

File: dockers/docker-teamd/start.sh [hunk 8]
```diff
@@ -6,7 +6,12 @@ rm -rf $TEAMD_CONF_PATH
 mkdir -p $TEAMD_CONF_PATH
 
 SONIC_ASIC_TYPE=$(sonic-cfggen -y /etc/sonic/sonic_version.yml -v asic_type)
-MAC_ADDRESS=$(ip link show eth0 | grep ether | awk '{print $2}')
+
+if [ "$SONIC_ASIC_TYPE" == "mellanox" ]; then
+    MAC_ADDRESS=$(sonic-cfggen -d -v DEVICE_METADATA.localhost.mac)
```
Comment (pavel-shirshov) [inline dockers/docker-teamd/start.sh]: can we do `sudo decode-syseeprom -m` here?

File: dockers/docker-teamd/start.sh (hunk 5 above)
Comment (andriymoroz-mlnx) [reply dockers/docker-teamd/start.sh]: the case under "else" getting MAC address from the "ip link show eth0" is the old code and used for other platforms
I can change it to what you suggested, but I am not able to test it on non-Mellanox boxes...

File: files/build_templates/sonic_debian_extension.j2 (hunk 6 above)
Comment (andriymoroz-mlnx) [reply files/build_templates/sonic_debian_extension.j2]: to get it in build log, I think
seen it in some other place
suggest to remove tee?

File: src/sonic-config-engine/sonic_platform.py (hunk 7 above)
Comment (andriymoroz-mlnx) [reply src/sonic-config-engine/sonic_platform.py]: like above - not sure about other platforms

File: dockers/docker-teamd/start.sh (hunk 8 above)
Comment (andriymoroz-mlnx) [reply dockers/docker-teamd/start.sh]: unfortunately no, it is not available in the containers...

**Comment by qiluo-msft** [comment]
> 	git clone https://github.com/Mellanox/hw-mgmt/ -b $(MLNX_HW_MANAGEMENT_VERSION) hw-management

This is a suggestion. Since the code is owned by Mellanox, why not commit the patch there?

---
Refers to: platform/mellanox/hw-management/Makefile:9 in 048db4d. [](commit_id = 048db4d847e7b9d3789d0d2ef64c499877396d5a, deletion_comment = False)

File: slave.mk [hunk 9]
```diff
@@ -481,6 +481,8 @@ $(addprefix $(TARGET_PATH)/, $(SONIC_INSTALLERS)) : $(TARGET_PATH)/% : \
 	j2 -f env files/initramfs-tools/union-mount.j2 onie-image.conf > files/initramfs-tools/union-mount
 	j2 -f env files/initramfs-tools/arista-convertfs.j2 onie-image.conf > files/initramfs-tools/arista-convertfs
 
+	j2 files/build_templates/updategraph.service.j2 > updategraph.service
```
Comment (jleveque) [inline slave.mk]: Is this line necessary? The same operation is being done in sonic_debian_extension.j2 above and there the resulting file is getting copied to its proper final destination. Why is this file getting written to the repo root? It doesn't appear to be used anywhere.

**Pull request merged**